pysph/sph/wc/crksph.py

```diff
@@ -373,3 +384,468 @@ def loop(self, d_idx, s_idx, d_ai, d_gradai, d_cwij, d_bi, d_gradbi, s_ai,
+        rij = tmpri/tmprj
+
+        tmprij = min(1, 4*rij/(1 + rij)**2.0)
+        phiij = max(0, tmprij)
+
+        tmpxij = dot(XIJ, XIJ, d)
+        etai_scalar = sqrt(tmpxij)/hi
+        etaj_scalar = sqrt(tmpxij)/hj
+        etaij = min(etai_scalar, etaj_scalar)
+
+        if etaij < eta_crit:
+            tmpphi = (etaij - eta_crit)/eta_fold
+            phiij = phiij * exp(-tmpphi*tmpphi)
+
+        for alp in range(d):
+            s = 0.0
+            for bet in range(d):
+                s += (d_gradv[d*d*d_idx + d*alp + bet] +
+                      s_gradv[d*d*s_idx + d*alp + bet]) * XIJ[bet]
+            tmpdvxij[alp] = s
+
+        uijhat[0] = d_u[d_idx] - s_u[s_idx] - 0.5*phiij * tmpdvxij[0]
+        uijhat[1] = d_v[d_idx] - s_v[s_idx] - 0.5*phiij * tmpdvxij[1]
+        uijhat[2] = d_w[d_idx] - s_w[s_idx] - 0.5*phiij * tmpdvxij[2]
+
+        tmpmui = dot(uijhat, XIJ, d) / (tmpxij/hi + EPS * hi)
+        mui = min(0, tmpmui)
+
+        tmpmuj = dot(uijhat, XIJ, d) / (tmpxij/hi + EPS * hj)
+        muj = min(0, tmpmuj)
+
+        Qi = rhoi * (-Cl*ci*mui + Cq*mui*mui)
+        Qj = rhoj * (-Cl*cj*muj + Cq*muj*muj)
+
+        fac = -(0.5/mi) * Vi * Vj * (pi + pj + Qi + Qj)
+
+        d_au[d_idx] += fac * DWIJ[0]
+        d_av[d_idx] += fac * DWIJ[1]
+        d_aw[d_idx] += fac * DWIJ[2]
+
+
+class EnergyEquation(Equation):
+    r"""**Energy Equation**
+
+    From [CRKSPH2017], equation (66):
+
+    .. math::
+        \Delta u_{ij} = \frac{f_{ij}}{2}\left(v_j^{\alpha}(t) + v_j^{\alpha}(t
+        + \Delta t) - v_i^{\alpha}(t) - v_i^{\alpha}(t + \Delta t)\right)
+        \frac{Dv_{ij}^{\alpha}}{Dt}
+
+    .. math::
+        f_{ij} = \begin{cases}
+        1/2 &|s_i - s_j| = 0,\\
+        s_{\min} / (s_{\min} + s_{\max}) &\Delta u_{ij}\times(s_i - s_j) > 0\\
+        s_{\max} / (s_{\min} + s_{\max}) &\Delta u_{ij}\times(s_i - s_j) < 0\\
+        \end{cases}
+
+    .. math::
+        s_{\min} = \min(|s_i|, |s_j|)
+
+    .. math::
+        s_{\max} = \max(|s_i|, |s_j|)
+
+    .. math::
+        s_{i/j} = \frac{p_{i/j}}{\rho_{i/j}^\gamma}
+
+    see MomentumEquation for :math:`\frac{Dv_{ij}^{\alpha}}{Dt}`
+    """
+
+    def __init__(self, dest, sources, dim, gamma):
+        self.dim = dim
+        self.gamma = gamma
+        super(EnergyEquation, self).__init__(dest, sources)
+
+    def _get_helpers_(self):
+        return [dot]
+
+    def initialize(self, d_idx, d_ae):
+        d_ae[d_idx] = 0.0
+
+    def loop(self, d_idx, s_idx, d_au, d_av, d_aw, d_ae, s_au, s_av, s_aw,
+             d_u0, d_v0, d_w0, s_u0, s_v0, s_w0, d_u, d_v, d_w, s_u, s_v, s_w,
+             d_p, d_rho, s_p, s_rho):
+        gamma = self.gamma
+        d = declare('int')
+        d = self.dim
+        auij, delu = declare('matrix(3)', 2)
+        auij[0] = d_au[d_idx] - s_au[s_idx]
+        auij[1] = d_av[d_idx] - s_av[s_idx]
+        auij[2] = d_aw[d_idx] - s_aw[s_idx]
+
+        delu[0] = s_u0[s_idx] + s_u[s_idx] - d_u0[d_idx] - d_u[d_idx]
```

Comment:
This doesn't seem quite correct as d_u is really not updated yet, perhaps you should set `ui = s_u + a_u*dt`?

Reply:
this energy equation should be called after the velocities are updated in the integrator, and after this energy should be updated.